aten/src/ATen/FunctionalizeFallbackKernel.cpp

```diff
@@ -75,6 +81,23 @@ namespace {
   }
 }
 
+at::Tensor to_device_functionalize(const at::Tensor & self, at::Device device, at::ScalarType dtype, bool non_blocking, bool copy, c10::optional<at::MemoryFormat> memory_format) {
+  if (!at::functionalization::impl::isFunctionalTensor(self)) {
+    // If the input is not a functional tensor, we want to "lift" the output to be functional.
+    // See Note [Functionalization <> torch.Tensor constructor]
+    at::AutoDispatchSkipFunctionalize guard;
+    auto out = self.to(device, dtype, non_blocking, copy, memory_format);
+    return at::functionalization::impl::to_functional_tensor(std::move(out));
+  } else {
+    // otherwise, to.device is a composite op, so let it decompose normally
+    return at::native::to(self, device, dtype, non_blocking, copy, memory_format);
+  }
+}
+
 TORCH_LIBRARY_IMPL(_, Functionalize, m) {
   m.fallback(torch::CppFunction::makeFromBoxedFunction<&functionalizeFallback>());
 }
+
+TORCH_LIBRARY_IMPL(aten, Functionalize, m) {
+  m.impl("to.device", TORCH_FN(to_device_functionalize));
```

Comment:
I updated the PR to use the approach where we construct an ordinary empty tensor and "lift" into a wrapper with the `.to()` call.

That required some changes though, since `to()` is `CompositeImplicitAutograd` (and by default functionalization will decompose all composite ops). If the `.to()` call no-ops (which it will in most cases, if the output dtype + device match), then we would just never hit a functionalization kernel and do the lifting.

I also can't use the boxed fallback (needed to write a custom kernel), since in the normal case, we *want* `.to()` to decompose. `to()` is actually a "problematic" op for functionalization, since it may-or-may-not return a view depending on whether or not it no-ops.

Reply:
`.to()` seems like the wrong way to get to a functionalize tensor...

Reply:
It seems like the constraint that you posed for yourself was to not have to modify the tensor_new.cpp (besides excluding functionalize key) but I think that is too much of a handicap. Once the tensor has been moved to the correct device, I think you should be willing to call another *new* operator that does the lifting.

Reply:
That sounds reasonable - I have two more questions:

(1) Should the other stuff in this function (functorch, autograd, tracer key, python key) also be updated to do the "lifting" separately from the `.to()` call? (Well I don't think it really makes sense for autograd, and it would be BC-breaking for tracing to trace a new op. Technically BC-breaking for python too, although maybe that's a better end state?)

(2) Should the "lifting" be an aten operator? (I think the answer is yes if we want to be able to do the lifting differently for different transforms, although lifting doesn't really make sense for e.g. backends).

Reply:
> Should the other stuff in this function (functorch, autograd, tracer key, python key) also be updated to do the "lifting" separately from the .to() call?

What makes this confusing is that traditionally, lifting is a no-op (and it should continue to be a no-op for all the cases it can be). The second confusing thing is that lifting has to be relative to *something*. In this case, we lift to the ambient mode (if there was no mode, lifting wouldn't do anything--but you could also imagine a variant of lifting that lifts to match another tensor; see for example https://github.com/albanD/subclass_zoo/blob/a4a0052832edaf6fb538095894c6c61ca6d5ba6a/simple_functorch.py#L896-L911 )

But yes, I think it should be an aten operator, but it should do nothing most of the time.